src/librustc_mir/transform/promote_consts.rs

```diff
@@ -502,9 +502,33 @@ impl<'tcx> Validator<'_, 'tcx> {
     fn validate_place(&self, place: PlaceRef<'tcx>) -> Result<(), Unpromotable> {
         match place {
             PlaceRef { local, projection: [] } => self.validate_local(local),
-            PlaceRef { local: _, projection: [proj_base @ .., elem] } => {
+            PlaceRef { local, projection: [proj_base @ .., elem] } => {
                 match *elem {
-                    ProjectionElem::Deref | ProjectionElem::Downcast(..) => {
+                    ProjectionElem::Deref => {
```

Comment:
Theoretically this `place` may be `*some_local.field`, where the projection is `[.field, *]`. In that case simply grabbing the `local` and reading its value won't be correct. While `check_static_ptr` requires the constant to be just a pointer, `struct Foo { field: &'static i32 }` would also be represented as just a pointer. So I think there should be an additional `proj_base.is_empty()` check. Since currently no MIR can get built that contains something like that, you can just `assert!(proj_base.is_empty())` after the `check_static_ptr` check. Please leave a comment with the above reasoning there so we can figure out the logic again in the future.

Reply:
Using `assert!` may lead to unnecessary panicking in the compiler, so I opt for simply checking this case and preventing the promotion.

Reply:
Well... if the compiler is buggy, we'd like to know ;) Right now there's no way that assertion could ever trigger that I can think of.

Reply:
I see. I found a test case that fails the assertion. It is `ui/consts/static-raw-pointer-interning.rs`. For convenience, I just post it here, with simplification so that MIR is reasonably readable.
```rust
static FOO: Foo = Foo {
    field: &42 as *const i32,
};

struct Foo {
    field: *const i32,
}

unsafe impl Sync for Foo {}

fn main() {
    let _ = unsafe { *FOO.field };
}
```
And the produced MIR is the following.
```
// MIR for `main` 0 mir_map

fn main() -> () {
    let mut _0: ();                      // return place in scope 0 at example4.rs:11:11: 11:11
    let mut _1: i32;                     // in scope 0 at example4.rs:12:13: 12:34
    let mut _2: &Foo;                    // in scope 0 at example4.rs:12:23: 12:26
    scope 1 {
    }
    scope 2 {
    }

    bb0: {
        StorageLive(_1);                 // scope 0 at example4.rs:12:13: 12:34
        StorageLive(_2);                 // scope 2 at example4.rs:12:23: 12:26
        _2 = const {alloc0: &Foo};       // scope 2 at example4.rs:12:23: 12:26
                                         // ty::Const
                                         // + ty: &Foo
                                         // + val: Value(Scalar(alloc0))
                                         // mir::Constant
                                         // + span: example4.rs:12:23: 12:26
                                         // + literal: Const { ty: &Foo, val: Value(Scalar(alloc0)) }
        _1 = (*((*_2).0: *const i32));   // scope 2 at example4.rs:12:22: 12:32
        StorageDead(_2);                 // scope 0 at example4.rs:12:34: 12:35
        StorageDead(_1);                 // scope 0 at example4.rs:12:34: 12:35
        _0 = const ();                   // scope 0 at example4.rs:11:11: 13:2
                                         // ty::Const
                                         // + ty: ()
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: example4.rs:11:11: 13:2
                                         // + literal: Const { ty: (), val: Value(Scalar(<ZST>)) }
        goto -> bb2;                     // scope 0 at example4.rs:13:2: 13:2
    }

    bb1 (cleanup): {
        resume;                          // scope 0 at example4.rs:11:1: 13:2
    }

    bb2: {
        return;                          // scope 0 at example4.rs:13:2: 13:2
    }
}

alloc0 (static: FOO, size: 8, align: 8) {
    ╾───────alloc3────────╼                         │ ╾──────╼
}

alloc3 (size: 4, align: 4) {
    2a 00 00 00                                     │ *...
}
```
The statement of index `3` in the `bb0` block triggers the assertion. You may see that there is a nested record field access.

Reply:
Ah, ok. That makes sense actually, so... I guess we need an additional test for

```rust
pub struct Slice(&'static [i32]);

static CONTENT: i32 = 42;
pub static CONTENT_MAP: Slice = Slice(&[CONTENT]);

pub static FOO: (i32, i32) = (42, 43);
pub static CONTENT_MAP2: Slice = Slice(&[FOO.0]);
```

Reply:
In that case, I think we're testing the wrong projection. We're testing the last projection, but we need to check the first one. So you need a runtime check for `proj_base.is_empty()` as you described